src/julia.h

```diff
@@ -1099,19 +1109,35 @@ typedef enum {
 
 STATIC_INLINE int jl_is_vararg_type(jl_value_t *v)
 {
+    v = jl_unwrap_unionall(v);
     return (jl_is_datatype(v) &&
-            ((jl_datatype_t*)(v))->name == jl_vararg_type->name);
+            ((jl_datatype_t*)(v))->name == jl_vararg_typename);
+}
+
+STATIC_INLINE jl_value_t *jl_unwrap_vararg(jl_value_t *v)
+{
+    return jl_tparam0(jl_unwrap_unionall(v));
 }
 
 STATIC_INLINE jl_vararg_kind_t jl_vararg_kind(jl_value_t *v)
 {
     if (!jl_is_vararg_type(v))
         return JL_VARARG_NONE;
+    jl_tvar_t *v1=NULL, *v2=NULL;
+    if (jl_is_unionall(v)) {
+        v1 = ((jl_unionall_t*)v)->var;
+        v = ((jl_unionall_t*)v)->body;
+        if (jl_is_unionall(v)) {
+            v2 = ((jl_unionall_t*)v)->var;
+            v = ((jl_unionall_t*)v)->body;
+        }
+    }
+    assert(jl_is_datatype(v));
```

Comment:
You're right, but I feel like this should be restricted or normalized somehow. You always want either `Tuple{Vararg{Array{T} where T}}` or `Tuple{Vararg{Array{T}}} where T`, not `Tuple{Vararg{Array{T}} where T}`. Of course, this is all just an artifact of the weird representation of vararg types, which makes `Tuple{Vararg{T} where T}` unavoidable, even though it does not make sense: you can't union together `Vararg` types. (`Union{Vararg, ...}` is already an error.)